Initial yubikey backend keytype support

tomis007 opened the pull request:
Same PR for initial yubikey piv support with the intermediate commits removed.

I also removed:
* changes to the config parsing (will address in another pr)
* prompting for yubikey pin (right now uses the default PIN)

Hopefully this helps for reviewing / testing. I think I also addressed all your comments from the previous review. Thank you for your help!

I do think the prompting for the PIN is important to avoid needing to store the pin anywhere, so once we get this PR done I can rework and submit that one separately

Foxboron commented:
Please stop making new PRs :) Rebase and force-push the branch in-place is much easier instead of leaving 3 PRs around in the project.

tomis007 commented:
Sure, I'll do that going forward, I didn't know that worked in a github PR my bad!

tomis007 commented:
Would it be helpful if rebased and force pushed this commit PR over https://github.com/Foxboron/sbctl/pull/413? I can then use this one for the prompt support PR

Foxboron commented:
> Would it be helpful if rebased and force pushed this commit PR over https://github.com/Foxboron/sbctl/pull/413? I can then use this one for the prompt support PR

Nah, no we can work in this branch :)

Foxboron commented on backend/backend.go:
```diff
@@ -295,15 +299,24 @@ func GetKeyHierarchy(vfs afero.Fs, state *config.State) (*KeyHierarchy, error) {
+			return "", err
+		}
+		return YubikeyBackend, nil
+	} else {
+		block, _ := pem.Decode(b)
+		// TODO: Add TSS2 keys
+		switch block.Type {
+		case "PRIVATE KEY":
+			return FileBackend, nil
+		case "TSS2 PRIVATE KEY":
+			return TPMBackend, nil
+		default:
+			return "", fmt.Errorf("unknown file type: %s", block.Type)
+		}
 	}
```
We don't need the else block, as the `if` block will return.

Foxboron commented on backend/yubikey.go:
```diff
@@ -0,0 +1,279 @@
+		},
+	}
+
+	h := md5.New()
+	h.Write(x509.MarshalPKCS1PublicKey(conf.PubKeyInfo.PublicKey.(*rsa.PublicKey)))
```
This code is repeated, so lets make this into a function.

Foxboron commented on cmd/sbctl/create-keys.go:
```diff
@@ -100,6 +107,7 @@ func RunCreateKeys(state *config.State) error {
 
 func createKeysCmdFlags(cmd *cobra.Command) {
 	f := cmd.Flags()
+	f.BoolVar(&OverwriteYubikey, "overwrite", false, "overwrite existing key if it exists in the Yubikey Signature slot")
```
Lets namespace this under `--yk-*` so `--yk-overwrite`.

Foxboron commented on cmd/sbctl/create-keys.go:
```diff
@@ -33,7 +36,7 @@ var createKeysCmd = &cobra.Command{
 	if state.Config.Landlock {
 		lsm.RestrictAdditionalPaths(
-			landlock.RWDirs(filepath.Dir(filepath.Dir(filepath.Clean(state.Config.Keydir)))),
+			landlock.RWDirs(filepath.Dir(filepath.Clean(state.Config.Keydir))),
```
Mmm, why is this being changed?

tomis007 replied:
I changed it trying to debug landlock with the yubikey, the change isn't actually necessary I reverted it back to what it was originally

Foxboron commented on backend/yubikey.go:
```diff
@@ -0,0 +1,279 @@
+	PublicKey string      `json:"PublicKey"`
+}
+
+var YK *piv.YubiKey
```
Any reason we can't throw the piv connetion into the `Yubikey` struct?

tomis007 replied:
I tried to do that originally but I ended up needing to open up a piv connection and track piv state in `YubikeyFromBytes` during the signing workflow and it became more complicated. I ended up going with  the most simple approach to start

Foxboron replied:
But can't we just have a pointer in the struct and use `connectToYubikey()` as a struct method when the actual card is needed? 

Everything seems fairly self-contained at the moment?

tomis007 replied:
That approach works for creating keys, but the tricky part is in the Signer interface. the piv-go library only lets you have one handle to the yubikey open at once (https://github.com/go-piv/piv-go/blob/0383b0aa884b2b642e9e3446ea01ba22ccadc83a/v2/piv/piv.go#L107). When you  `defer yk-piv.Close()` after creating the handle in `Signer()` the signing operation happens after the function is returned and by then the handle is closed. If you leave the handle open you can't open a new handle to the yubikey for future operations.

I can make it work but I would likely need to change the Signer interface to return a function that gets called after signing operations to appropriately Close the handle. What do you think about that approach?

Foxboron reviewed (changes requested):
Thanks for splitting stuff up :) Makes easier to review this!

tomis007 commented:
Updated with the requested changes, I also verified that using sbctl to create a key with the yubikey and signing files is working (for me at least) with the default pin

Foxboron commented on backend/yubikey.go:
```diff
@@ -0,0 +1,282 @@
+func (f *Yubikey) Certificate() *x509.Certificate { return f.cert }
+
+func (f *Yubikey) Signer() crypto.Signer {
+	auth := piv.KeyAuth{PIN: piv.DefaultPIN}
```
```suggestion
	auth := piv.KeyAuth{PIN: piv.DefaultPIN}
	if pin, found := os.LookupEnv("SBCTL_YUBIKEY_PIN"); found {
		auth = piv.KeyAuth{PIN: pin}
	}
```

Something like this, so you can escape the default pin.

Foxboron commented on backend/yubikey.go:
```diff
@@ -0,0 +1,282 @@
+//	   "PublicKey": publicKey
+//	}
+func (f *Yubikey) PrivateKeyBytes() []byte {
+	yubiData := YubikeyData{
+		Info:      *f.pubKeyInfo,
+		Slot:      "signature",
+		PublicKey: base64.StdEncoding.EncodeToString(x509.MarshalPKCS1PublicKey(f.pubKeyInfo.PublicKey.(*rsa.PublicKey))),
```
We same `f.pubKeyInfo` here, why do we have `YubiConfig.PubKeyInfo` as well stored as part of the state?

Is this to prevent us from reading the information from the yubikey multiple times?

tomis007 replied:
It's also stored in the config to make sure when you want to overwrite or create a new one, it only creates a new one on signature slot in the yubikey once

Foxboron commented:
I saw down to try this change and I can't seem to get it to work. The yubikey blinks for a press, but nothing happens when I press. I suspect the issue is on my side, but I found this a bit weird. Have to spend a little bit of time to figure out why this happens and the potential failure case.

tomis007 commented:
Interesting, I have not encountered that behavior before with my yubikey.

Are you able to create any key on your yubikey in the piv slot with `ykman`? 
```bash
ykman piv info
# you can set the touch/pin policy too
ykman piv keys generate --algorithm RSA2048 9a pubkey.test
```
What happens if you update yubikey.go and change the PIN/Touch policy to never?:
```go
piv.Key{
    Algorithm:   piv.AlgorithmRSA2048,
    PINPolicy:   piv.PINPolicyNever,
    TouchPolicy: piv.TouchPolicyNever,
}
```

tomis007 commented:
I just tested after resetting my yubikey `piv` data and it worked as expected:
```
$ ykman info
Device type: YubiKey 5C
Serial number: 12341234
Firmware version: 5.7.1
Form factor: Keychain (USB-C)
Enabled USB interfaces: OTP, FIDO, CCID

Applications
Yubico OTP      Enabled
FIDO U2F        Enabled
FIDO2           Enabled
OATH            Enabled
PIV             Enabled
OpenPGP         Enabled
YubiHSM Auth    Enabled

$ ykman piv reset
WARNING! This will delete all stored PIV data and restore factory settings. Proceed? [y/N]: y
Resetting PIV data...
Reset complete. All PIV data has been cleared from the YubiKey.
Your YubiKey now has the default PIN, PUK and Management Key:
        PIN:    123456
        PUK:    12345678
        Management Key: 010203040506070801020304050607080102030405060708


$ ./sbctl create-keys --keytype yubikey
Created Owner UUID 865d27c1-27a4-4b1d-bc5b-36ebf1d39124
Creating RSA2048 key...
Please press Yubikey to confirm presense
Created RSA2048 key MD5: 9a50e63ad7c616fb6aa8f0f5b6ebb680
Creating PK key...
Please press Yubikey to confirm presence for key RSA2048 MD5: 9a50e63ad7c616fb6aa8f0f5b6ebb680
Creating KEK key...
Please press Yubikey to confirm presence for key RSA2048 MD5: 9a50e63ad7c616fb6aa8f0f5b6ebb680
Creating db key...
Please press Yubikey to confirm presence for key RSA2048 MD5: 9a50e63ad7c616fb6aa8f0f5b6ebb680
✓
Secure boot keys created!
```

How far did it get when you tested? I wonder if there is a missing dependency, do you have `pcsclite` installed?

tomis007 commented:
Any luck testing? If you were unable to get it working I can take a look on a few different computers I have and try to reproduce the issue.

Foxboron commented:
Sorry!

I haven't managed to look at this again but I'll take a stab at this within the next couple of days :) Thanks!

Foxboron commented:
Hmmmmm, ran the same commands as you and reset the yubikey.

> couldn't initialize secure boot: command failed: smart card error 6d00

I suspect there is something wrong with scdaemon when it comes to exclusive vs shared access? I'll try and figure this out on my end

tomis007 commented:
Hmm, I'll do some trouble shooting as well. What yubikey / firmware version are you using?

Foxboron commented:
```
Device type: YubiKey 5 NFC
Serial number: 9027747
Firmware version: 5.1.1
Form factor: Keychain (USB-A)
Enabled USB interfaces: OTP, FIDO, CCID
NFC transport is enabled

Applications	USB          	NFC
Yubico OTP  	Enabled      	Enabled
FIDO U2F    	Enabled      	Enabled
FIDO2       	Enabled      	Enabled
OATH        	Enabled      	Enabled
PIV         	Enabled      	Disabled
OpenPGP     	Enabled      	Enabled
YubiHSM Auth	Not available	Not available
```

The one I tried with

tomis007 commented:
It seems like Yubico does not support upgrading/downgrading yubikey firmware, I haven't been able to trace down or reproduce your issue yet. Does your yubikey work as expected with `ykman`?

```
ykman piv info
ykman piv keys generate --algorithm RSA2048 9a pubkey.test
```

Foxboron commented:
I'll try and pick this up again. I do really want yubikey support solved for `sbctl` but I suspect I need to set off a bit more time to figure out why things are not working as expected.

tomis007 commented:
I tried with a different Yubikey (5 USB A) to make sure everything was working on the branch still:
```
$ git clone git@github.com:tomis007/sbctl.git -b initial-yubikey-support
$ make
a2x --no-xmllint --asciidoc-opts="-f docs/asciidoc.conf" -d manpage -f manpage -D docs docs/sbctl.8.txt
a2x --no-xmllint --asciidoc-opts="-f docs/asciidoc.conf" -d manpage -f manpage -D docs docs/sbctl.conf.5.txt
fatal: No names found, cannot describe anything.
go build -ldflags="-X github.com/foxboron/sbctl.Version=" -o sbctl ./cmd/sbctl
$ sudo ./sbctl create-keys --keytype yubikey
Created Owner UUID 018ba2df-8734-4889-bcb5-43f3a52fc21f
Creating RSA2048 key...
Please press Yubikey to confirm presense
Created RSA2048 key MD5: 99bec93ddf3d20fcc208583b611700b2
Creating PK key...
Please press Yubikey to confirm presence for key RSA2048 MD5: 99bec93ddf3d20fcc208583b611700b2
Creating KEK key...
Please press Yubikey to confirm presence for key RSA2048 MD5: 99bec93ddf3d20fcc208583b611700b2
Creating db key...
Please press Yubikey to confirm presence for key RSA2048 MD5: 99bec93ddf3d20fcc208583b611700b2
✓
Secure boot keys created!
$ ykman info
Device type: YubiKey 5 NFC
Serial number: 52734083
Firmware version: 5.7.4
Form factor: Keychain (USB-A)
Enabled USB interfaces: OTP, FIDO, CCID
NFC transport is enabled

Applications    USB     NFC
Yubico OTP      Enabled Enabled
FIDO U2F        Enabled Enabled
FIDO2           Enabled Enabled
OATH            Enabled Enabled
PIV             Enabled Enabled
OpenPGP         Enabled Enabled
YubiHSM Auth    Enabled Enabled
```

This was the same error that you described but your Yubikey seems newer than the ones in the issue: https://github.com/go-piv/piv-go/issues/55. I haven't been able to reproduce the issue unfortunately and I have no older yubikeys to test against

Foxboron commented:
Judging by https://github.com/FiloSottile/yubikey-agent/tree/main I'm wondering if we should omit `KeyInfo` and just save the public key and make assumptions about key policies and so on. Would that work?

We could alternatively store it if the yubikey is of a recent version.

tomis007 commented:
Oh nice find on the cause of the issue. I think it'd be ideal to ensure it works on older firmware versions too rather than just supporting newer ones. I'll take a look at how yubikey-agent manages older keys. It might take me a few days to get around to updating the pr

tomis007 commented:
@Foxboron I updated and refactored to remove the KeyInfo calls so it should hopefully work with your yubikey now. I did some more refactoring to improve the code and the yubikey connection handling. It is ready for a review/testing

Foxboron commented on backend/backend.go:
```diff
@@ -295,6 +300,14 @@ func GetKeyHierarchy(vfs afero.Fs, state *config.State) (*KeyHierarchy, error) {
 
 func GetBackendType(b []byte) (BackendType, error) {
+	if json.Valid(b) {
+		var yubiData YubikeyData
+		if err := json.Unmarshal(b, &yubiData); err != nil {
```
You can probably just omit the variable assignment if the intention is to just check if the format serializes into `YubikeyData`.

> if err := json.Unmarshal(b, &YbuikeyData{}); err != nil {

Foxboron commented on backend/yubikey.go:
```diff
@@ -0,0 +1,218 @@
+	cert, err := yubikeyReader.GetPIVKeyCert()
+	if err != nil {
+		if !errors.Is(err, piv.ErrNotFound) {
+			return nil, err
```
Lets wrap this error!
```suggestion
			return nil, fmt.Errorf("failed finding yubikey: %v", err)
```

lespea replied:
Should use `%w` not `%v` when wrapping errors

Foxboron replied:
@lespea only if you want the caller to depend on the error "type". And in this case we don't know yet.

Foxboron commented on backend/yubikey.go:
```diff
@@ -0,0 +1,218 @@
+		switch yubiPub := cert.PublicKey.(type) {
+		case *rsa.PublicKey:
+			// RSA2048 Public Key
+			if yubiPub.N.BitLen() == 2048 && !yubikeyReader.Overwrite {
```
Mm, we hardcode `2048` RSA keys, but `sbctl` explicitly hardcodes `4096`.

Should we go for `4096` instead? I believe all yubikeys from 4 and onwards support this just fine.

Foxboron commented on backend/yubikey.go:
```diff
@@ -0,0 +1,218 @@
+	var yubiData YubikeyData
+	err := json.Unmarshal(keyb, &yubiData)
+	if err != nil {
+		logging.Errorf("Error unmarshalling Yubikey: %v\n", err)
+		return nil, err
```
Return this as an error instead of `logging.Errorf` I think.

Foxboron commented on config/config.go:
```diff
@@ -133,12 +139,106 @@ func NewConfig(b []byte) (*Config, error) {
 	return conf, nil
 }
 
+// A type to wrap piv.Yubikey to manage the yubikey handle
+type YubikeyReader struct {
+	key       *piv.YubiKey
+	Overwrite bool
+}
```
Lets throw this into `config/yubikeyreader.go` instead of `config/config.go`

Foxboron commented on config/config.go:
```diff
@@ -133,12 +139,106 @@ func NewConfig(b []byte) (*Config, error) {
+}
+
+func (y *YubikeyReader) connectToYubikey() error {
+	if y.key != nil {
+		return nil
+	}
```
As we are just returning `nil` here, we can remove all the `if y.key == nil` checks earlier in the code.

Foxboron commented on lsm/lsm.go:
```diff
@@ -41,6 +41,8 @@ func LandlockRulesFromConfig(conf *config.Config) {
 			conf.GUID,
 			conf.FilesDb,
 			conf.BundlesDb,
+			//TODO will this be different for others for yubikeys?
+			"/usr/lib/libpcsclite_real.so.1",
```
I think we'll just commit this as-is and we'll figure out where it doesn't work in the future when we get bug reports :)

tomis007 replied:
Agreed

Foxboron commented:
Generally I think this is shaping up nicely. I'm a bit concerned as there is not a good way to test this at all. So I might merge this and see if I can find some time to check if I can do a small test setup before release.

Else I'll be a little bit dependent on you being willing to help debug issues :)

tomis007 commented on config/config.go:
```diff
@@ -133,12 +139,106 @@ func NewConfig(b []byte) (*Config, error) {
+			return fmt.Errorf("timeout waiting for yubikey")
+		case cards = <-c:
+		}
+	}
```
I think this pattern is fairly common in go (https://gobyexample.com/timeouts / https://go.dev/blog/concurrency-timeouts). I did refactor it to ensure that the goroutine returns when the timeout finishes (before the goroutine kept polling while the program exited)

I don't think the 
```
for {
    select {
    ...
    }
}
```
structure works because the select timeout is reset after each 500 Millisecond and it doesn't end up timing out. I think I might just be misunderstanding though how you want to refactor it. I think you need a goroutine to do the polling with the timeout.

tomis007 replied:
Ah yeah that makes sense if the goal is to get rid of the goroutine. I updated it to poll without the go call()

tomis007 commented:
I added all the requested changes, except the polling on the yubikey. I am happy to help with future issues! Let me know what you think on the updated code

Were you able to create / sign a uki with your yubikey?

One thing that I am tracking is that I think if the pin is non default there may be an issue with the hook after a uki rebuild gets triggered. I think that's a future issue to figure out though as this pr as already quite large

Foxboron commented:
>  I think that's a future issue to figure out though as this pr is already quite large and eventually I'd like to be able to prompt for the pin instead of storing it

Yes. If we just keep this to the basic yubikey functionality and then try to expend with pin reading later it will be much simpler for us :)

The entire process of prompting for information when we are trying to run interactively is probably going to be tricky. Which was my primary motivation for splitting this up.

tomis007 commented:
Requested changes are addressed and I confirmed yubikey creation/signing is working for my two yubikeys (both are on 5.7.X)

tomis007 commented:
Sure, I'll clean it up in the next day or two, thanks for all the help!

tomis007 commented:
I cleaned up the commit history. I thought about breaking it into a few different commits, but the yubikey support did not have a clean breakdown and the commits in between were not really functional and in a buggy/incomplete state. I ended up just rebaseing to the current state and fixuped the yubikey specific changes.

Foxboron commented:
I think that is completely fair :)

I'll do a last review and test, hopefully tonight.

Foxboron commented on README.md:
````diff
@@ -261,3 +261,32 @@ Enrolled bundles:
 ```
+
+
+## Yubikey Support
+To create a signing key with a yubikey, use the `--keytype yubikey` flag with the `create-key` command. For example:
+
````
Much appreciated, this also needs to go into the man page, but I can do that for you if you want?

tomis007 replied:
Sure that'd be great if you want to add to the man page. I did notice in the man page a few other options are missing too: like for create keys a few options aren't  specified:
```
**create-keys**::
        Creates a set of signing keys used to sign EFI binaries. Currently, it
        will create the following keys:
        * Platform Key
        * Key Exchange Key
        * Signature Database Key

        *-e*, *--export*;;
                The directory to persist the exported keys.

        *-d*, *--database-path*;;
                Path to save the GUID file when generating keys.
```
vs
```
sbctl create-keys -h
Create a set of secure boot signing keys

Usage:
  sbctl create-keys [flags]

Flags:
  -d, --database-path string   location to create GUID file
      --db-keytype string      db key type (default: file)
  -e, --export string          export file path
  -h, --help                   help for create-keys
      --kek-keytype string     KEK key type (default: file)
      --keytype string         key type for all keys
      --pk-keytype string      PK key type (default: file)
      --yk-overwrite           overwrite existing key if it exists in the Yubikey Signature slot

Global Flags:
      --config string      Path to configuration file
      --debug              Enable verbose debug logging
      --disable-landlock   Disable landlock sandboxing
      --json               Output as json
      --quiet              Mute info from logging
```

Foxboron replied:
I'll solve that then :) thanks!

Foxboron commented:
The CI does not like this. But I'll try and figure it out. It builds fine locally
